### PR TITLE
qubes 4.0rc4 and init refactoring

### PR DESCRIPTION
**Pull request by flammit**

### CODE REVIEW

**Comment by flammit** [comment]
Split out the KGPE-D16 support to another branch

**Comment by osresearch** [comment]
The qubes part of this PR is no longer relevant after fixing #227, right?

I'm hesitant on the split of `boards/$(BOARD).config` and `.mk` since it makes for more files that have to be tracked and possibly maintained in sync.  What do you think about a way to generate the `/etc/config` file from all of the `$(CONFIG_.*)` variables?

`export | grep '^CONFIG_'` almost works, except that most of these configs are not exported.

**Comment by flammit** [comment]
Yup the qubes isn't needed now and the export way to get `/etc/config` SGTM.  Let me know if you need me to fix.

**Comment by osresearch** [comment]
I've pushed a fix that generates `/etc/config` from the environment: https://github.com/osresearch/heads/commit/4f5432bb461ad63995afb45aba34534e1d8c3d23

What else needs to be merged from this PR?  I think the TPM fixes went in as part of merging #293

**Comment by flammit** [comment]
Just the `initrd/init` changes, but as long as it looks like #331 in the end, it should be fine to close / ignore this PR.

File: initrd/init [hunk 1]
```diff
@@ -16,50 +16,36 @@ mkdir /proc /sys /dev /tmp /boot /media 2>&- 1>&-
 mount /dev 2>/dev/ttyprintk
 mount /proc 2>/dev/ttyprintk
 mount /sys 2>/dev/ttyprintk
-mount /sys/firmware/efi/efivars
-
-# Setup the pty psudeo filesystem
-mkdir /dev/pts
```
Comment (osresearch) [inline initrd/init]: What's wrong with always mounting `/dev/pts`?

File: initrd/init [hunk 2]
```diff
@@ -69,35 +55,76 @@ if [ ! -z "$CONFIG_USB_BOOT_DEV" ]; then
 	echo >> /etc/fstab "$CONFIG_USB_BOOT_DEV /media auto defaults,ro 0 0"
 fi
 
-if [ ! -x "$CONFIG_BOOTSCRIPT" ]; then
-	recovery 'Boot script missing?  Entering recovery shell'
-	# just in case...
-	tpm extend -ix 4 -ic recovery
-	exec /bin/ash
+# Now it is safe to print a banner
+if [ ! -z "$CONFIG_LINUXBOOT" ]; then
```
Comment (osresearch) [inline initrd/init]: In my recent `export CONFIG_*` commit I didn't export the linuxboot config. Do you think it is important to be able to distinguish them?

Also, `[ ! -z "$CONFIG_LINUXBOOT" ]` will go down this path if `CONFIG_LINUXBOOT=n`.  It is better to test against `y`.

File: initrd/init [hunk 3]
```diff
@@ -69,35 +55,76 @@ if [ ! -z "$CONFIG_USB_BOOT_DEV" ]; then
...
-	boot_option
-echo
-
-if [ "$boot_option" = "r" ]; then
-	# Start an interactive shell
-	recovery 'User requested recovery shell'
-	# just in case...
-	tpm extend -ix 4 -ic recovery
-	exec /bin/ash
+# Setup remote attestation interface
```
Comment (osresearch) [inline initrd/init]: The network setup should move to a separate setup file, rather than living here in `/init`.

File: initrd/init (hunk 1 above)
Comment (flammit) [reply initrd/init]: Mmm... `CONFIG_UNIX98_PTYS=y` isn't always built in the kernel so mounting devpts can fail but I guess that's not awful (added `CONFIG_LINUX_PTY` in board config in PR #331).

File: initrd/init (hunk 3 above)
Comment (flammit) [reply initrd/init]: Fine by me - I just wanted minimal equal functionality changes to what you had done for linuxboot.

**Comment by flammit** [comment]
If it makes sense, I can just rebase/refactor #327, #330 and #331 to current master and incorporate all your comments.

**Pull request closed without merging**